io/native/src/main/scala/fs2/io/net/tls/TLSSocketPlatform.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+package io
+package net
+package tls
+
+import cats.effect.kernel.Async
+import cats.effect.kernel.Resource
+import cats.effect.std.Semaphore
+import cats.syntax.all._
+import com.comcast.ip4s.IpAddress
+import com.comcast.ip4s.SocketAddress
+
+private[tls] trait TLSSocketPlatform[F[_]]
+
+private[tls] trait TLSSocketCompanionPlatform { self: TLSSocket.type =>
+
+  private[tls] def apply[F[_]: Async](
+      socket: Socket[F],
+      connection: S2nConnection[F]
+  ): Resource[F, TLSSocket[F]] =
+    Resource.eval(mk(socket, connection)) <*
+      Resource.makeFull[F, Unit](poll => poll(connection.handshake))(_ =>
+        connection.shutdown.attempt.void
+      )
```

Comment:
The `.attempt` here is a bit unfortunate. It seems on the JVM `endOfOutput` may be called on the socket before closing the TLS connection, which makes graceful shutdown impossible. I assume this is incorrect use of APIs, but for sake of cross-compilability we suppress the error.